### PR TITLE
client side grpc implementation

### PR DESCRIPTION
**Pull request by vforvalerio87**
Allows client to make both grpc+json and grpc+proto calls to the daemon

Minor fix: no more requests for the list of featured services once it has been fetched

### CODE REVIEW

File: src/components/job.js [hunk 1]
```diff
@@ -162,29 +165,40 @@ class Job extends React.Component {
...
+          const requestObject = params
 
+          const serviceSpecJSON = Root.fromJSON(this.props.serviceSpec)
+          const packageName = Object.keys(serviceSpecJSON.nested)
+            .find(key =>
+              typeof serviceSpecJSON.nested[key] === "object" &&
+              serviceSpecJSON.nested[key].hasOwnProperty("nested") &&
+              typeof serviceSpecJSON.nested[key].nested !== "undefined"
+            )
+          const serviceName = typeof packageName !== "undefined" ?
```
[the remark this answers is not in the document]
Comment (vforvalerio87) [reply src/components/job.js]: Agreed. Checks for hasOwnProperty + whether it's undefined have been moved to a dedicated function. I'm moving everything to a dedicated Client class which will abstract the grpc+json vs grpc+proto calls, generate urls, generate the rpcImpl function from the factory, etc.

File: src/grpc.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    .then(buffer => {
+      return grpcJSONResponseToString(buffer)
+    })
+    .catch(console.error)
+}
+
+export function rpcImpl(host, packageName, serviceName, methodName, requestHeaders) {
+  return (method, requestObject, callback) => {
+    const service = [ packageName, serviceName ].filter(Boolean).join(".")
+    window.fetch(`${host}/${service}/methodName`, {
```
Comment (arturgontijo) [inline src/grpc.js]: I'm getting error here.
I think this should be: `window.fetch(`${host}/${service}/${methodName}`, {`

File: src/index.js [hunk 3]
```diff
@@ -140,14 +142,24 @@ class App extends React.Component {
...
     console.log("Agent " + agent.name + " selected");
-    
-    this.setState({
-      selectedAgent: agent,
-      serviceCallComponent: this.serviceNameToComponent[agent.name] || this.serviceDefaultComponent,
-      usingDefaultCallComponent: !(agent.name in this.serviceNameToComponent),
-    });
+    Promise.all([
+      window.fetch(`${agent.endpoint}/encoding`),
+      window.fetch(`http://protobufjs.singularitynet.io/${agent.address}`)
```
Comment (arturgontijo) [inline src/index.js]: I've tried with the service `SNET_BR/named-entity-recognition` and I've got:
`GET http://protobufjs.singularitynet.io/0x06b02f360e35f13781704a598ab85f32137c429b 400 (Bad Request)`

Comment (vforvalerio87) [reply src/index.js]: So, the error message you get associated with the 400 (Bad Request) is "Service spec has more than 1 service".

Here's the big picture:
.proto definitions are compiled by a server component called [service-spec-provider](https://github.com/singnet/service-spec-provider) to a .json format that protobufjs can use to encode/decode the proto requests/responses in the browser, and to provide the dapp with the package, service and method names.

The issue with .proto definitions having more than one service in them, for example sharing the same file or files for multiple services, is identifying which service is actually exposed at the endpoint specified in the Agent contract, either by the application or the user.

You could have one .proto file with 5 services, each of which is deployed at a different endpoint in different servers (or at different ports). So, the user would either see an select input, generated from the protobufjs .json, with the different services available in the source .proto file or files, or have to manually put in the name of a service to call, with no way to know from either the contract or the .json file itself if the services specified in the .proto definitions are actually available at the endpoint specified in the Agent contract.

I can think of two ways to go about this: either we agree on a convention to allow users to specify the name of the service (or services) offered registered with the specified endpoint in the smart contract itself and we communicate that to the users, or we allow developers to only have one service in each .proto definition. I would go with this second option. See [this comment](https://github.com/singnet/alpha-dapp/pull/33#issuecomment-431080179) for the original discussion.

**Comment by astroseger** [comment]
@vforvalerio87 
It seems that having multiply services in the single .proto file shoudn't be a problem. We only need to require that protobuf Service name is equal to Registry service name. (at least it is total fine with new MPE payment system, and I don't see why it is a problem with old Agent based system).
1. The client know the name of service he is using (proto-buf-service-name = Registry-service-name)
2. The client know the list of endpoints (in the case of Agent --- single endpoint), which corresponded to this service. So there is already well defined way for the client to know which endpoints actually support this service.
3. The daemon+server know which service he should work with :) 
Moreover we can have the situation in which single end-point support multiply services (it is not yet implemented in the daemon, but as I see it, it is quite straightforward to add.. )

**Comment by vforvalerio87** [comment]
It's one of the possibilities, and it's one I can agree on. It's something we'll need to agree on and communicate. In case we go with it, I think we might also want to accomodate the current software to facilitate the practice to help developers (example: when you upload the tar with the spec for the current service, the CLI should warn you if the service name in the registry for the current service does not match one of the service names in the .proto definition you are packaging).

Some of the services do not currently conform to that specific, so they will have to be accommodated. (Example: see [0xb57b4c70379e84cd8d42a867cf326d5e0743e11d](http://protobufjs.singularitynet.io/0xb57b4c70379e84cd8d42a867cf326d5e0743e11d) which has a ServiceName in the .proto definition of "NetowrkAnalytics" while the name in the Registry is "NetworkAnalyticsServices").

**Comment by astroseger** [comment]
I think having "protobuf service name" == "registry service name" will add consistency. It will solve the current problem, and probably will prevent some future problem as well. 
Hopefully next week, I will implement snet-cli functionality for uploading registry-related metadata (=publishing service in new MPE payment system). So, I will forbid to upload .proto file which doesn't have a service with "Registry" name.
I also will add such prohibition (upload .proto file without correct service) to old functionality (uploading .proto file in the case of old payment system).

**Comment by vforvalerio87** [comment]
Ok cool I'll do the dapp fix too then. Also, I would drop the old registry from the dapp

**Comment by arturgontijo** [comment]
I've put a python-flask dapp online just to get a GUI for our office's services.

What I'm doing is:
1 - Flask is running on process 0 (P[0]).
2 - A Registry "getter" on P[1]. This guy gets the list of organizations/services, compile the .proto and creates a JSON with the services specs to be used by P[0] to render the frontend. P[1] keeps running in an infinite loop (updating).
3 - The last part is a python script (P[2]) to call the service using as arguments the `spec_ipfs_hash`, `endpoint`, `method`, `params`, `job_address` and `job_signature`. I have to do this way to avoid the naming collision at the Protobuf's DescriptorPool (that appears when you import a grpc code-gen that has same variable name of a previous imported module). The script runs and exits after gets the service's response, freeing the DesciptorPool.

The app is running at http://54.203.198.53:7001/
This is a POC so the code is poor and the frontend is awful.
This kind of stack will never be a final solution but maybe we can use some of its idea to get our beta-dapp up.

To test it, you can call the SNET_BH/BasicService.
SNET_BH/ObjectDetection: `model: yolov3`, `img_path: link`, `confidence: 0.1`.
SNET_BH/ImageRecon: Fowers|Dogs `model: ResNet152`, `img_path: link`.
MetaMask must be enabled.

PS: Right now I'm trying to download the .proto only at P[2] not at P[1] anymore.

**Comment by vforvalerio87** [comment]
How do you associate a grpc Service with a specific service object in the JSON though when you have multiple services in the same tar, possibly in different .proto files?

I pushed the change to associate the service names to the name of the registered service in the Registry and in most service specs that have been uploaded to the Registry they do not match.

**Comment by arturgontijo** [comment]
The json from ImageRecon (that has 3 service) is:
```
"SNET_BH": {
	"ImageRecon": {
		"path": "", 
		"agent_address": "0x9211Ca9A96063401BaAC5cafE85efC4024325279",
		"tags": ["ResNet", "AlexNet", "InceptionV3", "Image Recognition", "Flowers", "Dogs"],
		"price": 2,
		"endpoint": "http://54.203.198.53:7007",
		"metadata_uri": "ipfs:/ipfs/QmXYo1vm1GapybM7ggjhS1wbJJMe3CL3S7hrNaT8jaymfk",
		"spec_hash": ["QmbjCY5X1DwHppbhEjgDzHkvK5RJqFKmBWrbfqQnRMg2BT"],
		"service_spec": {
			"Flowers": {
				"method": "flowers",
				"request": {
					"ImageReconRequest": {
						"params": {
							"model": {"type": 9},
							"img_path": {"type": 9}}}},
				"response": {
					"Result": {
						"params": {
							"delta_time": {"type": 9},
							"top_5": {"type": 9}}}}},
			"Dogs": {
				"method": "dogs",
				"request": {
					"ImageReconRequest": {
						"params": {
							"model": {"type": 9},
							"img_path": {"type": 9}}}},
				"response": {
					"Result": {
						"params": {
							"delta_time": {"type": 9},
							"top_5": {"type": 9}}}}},
			"Cars": {
				...}}}}
```
What I'm doing is like a snet-cli GUI, my call:
https://github.com/arturgontijo/flask-grpc/blob/master/backend/service_utils/__init__.py#L76

I didn't test it with services that have more than one .proto.
I'll publish one here and test it.

**Comment by vforvalerio87** [comment]
Nice work but there are a few issues here.
The call isn't actually made from the browser as it is with the current dapp, and how it would have to be for a browser SDK. The way PR works, the dapp makes the call from the browser itself, which is the point of grpc-web.
Also, the link you posted doesn't work if the encoding is not .proto, which is the case for JSON rpc backends. I'm not talking about the request from the client to the daemon but the request from the daemon to the service backend. The dapp queries the "encoding" endpoint of the daemon, properly frames grpc+json requests and makes the calls.

The dapp also works if the service uses more than one .proto file. As long as no file has conflicts in terms of packageName/serviceName/methodName (example: methodName is unique) the dapp will be able to make the call.

However, for browser calls made from the dapp to the services to work, you must use a version of the daemon that allows CORS requests for unregistered service endpoints (see [#snet-daemon/pull/65](https://github.com/singnet/snet-daemon/pull/65))

**Comment by arturgontijo** [comment]
Agree!
My solution is just to test some of our services that use image (it's temporary).

I know that the beta-dapp must be able to call a service from browser and I also know that this is kind of annoying to get done.

About the "encoding", I removed it from my call() because I thought that wasn't necessary anymore.
But I can put it back, it comes from snet-cli `commands.py` too:
https://github.com/singnet/snet-cli/blob/master/snet_cli/commands.py#L696-L708

I'll keep testing your fork and give you some feedbacks. Thanks!

**Comment by vforvalerio87** [comment]
@arturgontijo could you try updating the daemon on your server? The service should work with no problem from the browser. You'd just have to fill the fields by hand as I still have to finish the auto generated UI (feel free to contribute)

Example (for SNET_BH/BasicService):
Service name: Multiplication
Method name: mul
Params: {"a": 1, "b": 3}

**Comment by arturgontijo** [comment]
> @arturgontijo could you try updating the daemon on your server? The service should work with no problem from the browser. You'd just have to fill the fields by hand as I still have to finish the auto generated UI (feel free to contribute)
> 
> Example (for SNET_BH/BasicService):
> Service name: Multiplication
> Method name: mul
> Params: {"a": 1, "b": 3}

Now `BasicService` is running with the updated `snet-daemon` .
I tested it with `snet client call ...` and it's working.
With my flask-dapp I'm getting error when confirming transaction using MetaMask.
I'll clone your alpha-dapp and test it via browser too.

**Comment by astroseger** [comment]
@vforvalerio87 Yes, it makes sense.
So, the full name of the method in gRPC mode is actually protobuf_service_name+protobuf_method_name. 

My implementation of call in "snet-cli mpe-client"  already support it (you should specify both protobuf_service_name and protobuf_method_name). 

But old snet-cli implementation of call using jobs doesn't support it! But I think it is fine, because I believe it will become obsolete rather soon.

**Comment by vforvalerio87** [comment]
@astroseger Actually it's more complicated than that.

gRPC messages can belong to a package, so by default the address is:
"<endpoint>/<package name of the service>.<service name>/<method>"

@arturgontijo you can actually try with the [live one](http://alpha.singularitynet.io/).
Since it was broken anyways I deployed the version from this PR. I tested it with both your grpc to grpc service SNET_BH/BasicService and with the grpc to json SingularityNET/ExampleService

You can view the different service specs in the JSON format the dapp uses at http://protobufjs.singularitynet.io/<agent_address> ([example](http://protobufjs.singularitynet.io/0xf2b35821409baee26afcc955d7a332031003d95d)) or http://protobufjs.singularitynet.io/<org_name>/<service_name> ([example](http://protobufjs.singularitynet.io/SNET_BH/BasicService)) to see which services/methods are there for each agent and what parameters they take

**Comment by astroseger** [comment]
@vforvalerio87  Should we support multiply .proto files. As I see now in the Agent we store .tar, not simple one .proto file. But it seems having multiply protofiles is overkill (as well as having several packageNames). How we are going price all of them?

**Comment by vforvalerio87** [comment]
@astroseger I'm supporting it because you might have a scenario where a .proto file, which has a service for example, imports other .proto files which have the messages, or types, for that service.
It comes from two considerations:
1) Allow people to upload definitions for pre-existing gRPC services they already have and use for other purposes (example: their own SaaS) without having to manually modify them to work with SingularityNET every time they change them
2) Eventually, we would like to support Type repositories (already included in the Registry) to allow organizations to upload shared types which other services can use and reference. This is especially important for services to be able to dynamically call other services in our ecosystem

In the [service-spec-provider](https://github.com/singnet/service-spec-provider) I'm including an endpoint which serves [a service spec compiled from multiple files which also includes a package](http://protobufjs.singularitynet.io/test). This spec though has only one service. If it had multiple services and they were all in the same file, the JSON would not change. If it had multiple services spread amongst different files, the top-level JSON array would include more objects (one per top-level file with a service in it).

**Pull request merged**